h2o-core/src/main/java/water/parser/ParseTime.java

```diff
@@ -73,19 +74,29 @@ private static long attemptYearFirstTimeParse(BufferedString str) {
     yyyy = digit(yyyy,buf[i++]);
     yyyy = digit(yyyy,buf[i++]);
     yyyy = digit(yyyy,buf[i++]);
-    if( buf[i++] != '-' ) return Long.MIN_VALUE;
-    MM = digit(MM,buf[i++]);
-    MM = i<end && buf[i]!='-' ? digit(MM,buf[i++]) : MM;
-    if( MM < 1 || MM > 12 ) return Long.MIN_VALUE;
-    if( (end-i)>=2 ) {
-      if( buf[i++] != '-' ) return Long.MIN_VALUE;
-      dd = digit(dd, buf[i++]);
-      dd = i < end && buf[i] >= '0' && buf[i] <= '9' ? digit(dd, buf[i++]) : dd;
-      if( dd < 1 || dd > 31 ) return Long.MIN_VALUE;
-    } else dd=1; // no day
-    while( i < end && buf[i] == ' ' ) i++;
-    if( i==end )
-      return new DateTime(yyyy,MM,dd,0,0,0, getTimezone()).getMillis();
+    if( buf[i] != '-' ) {  // YYYMM[dd]
+      MM = digit(MM, buf[i++]);
+      MM = i < end ? digit(MM, buf[i++]) : MM;
+    } else {
+      if (buf[i++] != '-') return Long.MIN_VALUE;
+      MM = digit(MM, buf[i++]);
+      MM = i < end && buf[i] != '-' ? digit(MM, buf[i++]) : MM;
+    }
+    if (MM < 1 || MM > 12) return Long.MIN_VALUE;  // invalid 2digit month
+    if ((end - i) >= 2) {
+      if( buf[i]!= '-' ) {
+        dd = digit(dd, buf[i++]);
+        dd = i < end && (buf[i] >= '0' && buf[i] <= '9') ? digit(dd, buf[i++]) : dd;
+      }else {
+        if (buf[i++] != '-') return Long.MIN_VALUE;
+        dd = digit(dd, buf[i++]);
+        dd = i < end && buf[i] >= '0' && buf[i] <= '9' ? digit(dd, buf[i++]) : dd;
+      }
+      if (dd < 1 || dd > 31) return Long.MIN_VALUE;  //invalid day
+    } else dd = 1; // no day
+    while (i < end && (buf[i] == ' ' || buf[i] == '-')) i++;
+    if (i == end)
+      return new DateTime(yyyy, MM, dd, 0, 0, 0, getTimezone()).getMillis();
```

Comment:
Wikipedia mentioned, that dates before 1583 are not automatically allowed (https://en.wikipedia.org/wiki/ISO_8601). Do we want to put a warning into logs  here?
It would help debugging...